### PR TITLE
Implement the @@unscopables well-known Symbol

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 1]
```diff
@@ -1489,6 +1489,22 @@ static Value builtinArrayShift(ExecutionState& state, Value thisValue, size_t ar
     return first;
 }
 
+static Value builtinArrayUnscopables(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
+{
+  Value blacklist = state.context()->globalObject()->objectCreate()->call(state, Value(), 1, &blacklist);
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Isn't it too complicated way to create an empty object?

File: src/runtime/EnvironmentRecord.h [hunk 2]
```diff
@@ -225,9 +228,20 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
...
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(atomicName));
-        if (result.hasValue()) {
-            return BindingSlot(this, SIZE_MAX - 1);
-        } else {
+        Object* bindings = result.value(state, Value(0)).asObject();
+        if (result.hasValue() == true) {
+          Value unscopables = bindings->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, Value(0));
+          if(unscopables.isObject() == true)
+          {
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Move the opening brace next to the condition.
```
if(unscopables.isObject() == true) {
```

File: src/runtime/EnvironmentRecord.h [hunk 3]
```diff
@@ -225,9 +228,20 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
...
+            bool blocked = unscopables.asObject()->get(state, ObjectPropertyName(state, atomicName)).value(state, Value(0)).asBoolean();
+            if(blocked == true)
+            {
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Ditto.

File: src/runtime/EnvironmentRecord.h [hunk 4]
```diff
@@ -225,9 +228,20 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
...
+              return BindingSlot(this, SIZE_MAX);
+            }
+          }
+          return BindingSlot(this, SIZE_MAX-1);
+        }
+         else {
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Misaligned line.

File: src/runtime/GlobalObjectBuiltinArray.cpp (hunk 1 above)
Comment (rerobika) [reply src/runtime/GlobalObjectBuiltinArray.cpp]: Otherwise, the approach is incorrect, since according to the standard the blackList object's prototype should be null.

```
Object * blackList = new Object(state);
blackList->setPrototype(state, nullptr);
```

After that all `blackList.asObject()` casts become unnecessary.

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 5]
```diff
@@ -1489,6 +1489,22 @@ static Value builtinArrayShift(ExecutionState& state, Value thisValue, size_t ar
...
+  blacklist.asObject()->defineOwnPropertyThrowsException(state, ObjectPropertyName(state.context()->staticStrings().copyWithin), ObjectPropertyDescriptor(Value(true)));
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: I think you can use `defineOwnProperty` method here, since `blackList` is an intact object, so it cannot throw any exception during the property set operation.

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 6]
```diff
@@ -1808,6 +1824,8 @@ void GlobalObject::installArray(ExecutionState& state)
...
     m_arrayPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state.context()->staticStrings().keys),
                                                        ObjectPropertyDescriptor(new FunctionObject(state, NativeFunctionInfo(state.context()->staticStrings().keys, builtinArrayKeys, 0, nullptr, NativeFunctionInfo::Strict)), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::WritablePresent | ObjectPropertyDescriptor::ConfigurablePresent)));
+    m_arrayPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state.context()->staticStrings().unscopables),
+                                                       ObjectPropertyDescriptor(new FunctionObject(state, NativeFunctionInfo(state.context()->staticStrings().getSymbolUnscopables, builtinArrayUnscopables, 0, nullptr, NativeFunctionInfo::Strict)), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::WritablePresent | ObjectPropertyDescriptor::ConfigurablePresent)));
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: `This property has the attributes { [[Writable]]: false, [[Enumerable]]: false, [[Configurable]]: true }.`

File: src/runtime/EnvironmentRecord.h [hunk 7]
```diff
@@ -216,6 +219,12 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(name));
         if (result.hasValue()) {
+            Value unscopables = m_bindingObject->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, m_bindingObject);
+            if (unscopables.isObject() == true) {
+                if (unscopables.asObject()->get(state, ObjectPropertyName(state, name)).value(state, unscopables).toBoolean(state) == true) {
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Please merge these two if statements.

File: src/runtime/EnvironmentRecord.h [hunk 8]
```diff
@@ -226,6 +235,12 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(atomicName));
         if (result.hasValue()) {
+            Value unscopables = m_bindingObject->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, m_bindingObject);
+            if (unscopables.isObject() == true) {
+                if (unscopables.asObject()->get(state, ObjectPropertyName(state, atomicName)).value(state, unscopables).toBoolean(state) == true) {
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Ditto.

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 9]
```diff
@@ -1489,6 +1489,20 @@ static Value builtinArrayShift(ExecutionState& state, Value thisValue, size_t ar
     return first;
 }
 
+static Value builtinArrayUnscopables(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
+{
+    Object* blackList = new Object(state);
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: The `blackList` object prototype should be `null`.

Comment (kisbg) [reply src/runtime/GlobalObjectBuiltinArray.cpp]: @rerobika  Thank you for the reviews and for the suggestions!

**Comment by kisbg** [comment]
@rerobika  Thank you for the reviews and for the suggestions!

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 10]
```diff
@@ -1835,7 +1845,8 @@ void GlobalObject::installArray(ExecutionState& state)
                                                                ObjectPropertyDescriptor(new FunctionObject(state, NativeFunctionInfo(state.context()->staticStrings().next, builtinArrayIteratorNext, 0, nullptr, NativeFunctionInfo::Strict)), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::WritablePresent | ObjectPropertyDescriptor::ConfigurablePresent)));
     m_arrayIteratorPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state, Value(state.context()->vmInstance()->globalSymbols().toStringTag)),
                                                                ObjectPropertyDescriptor(Value(String::fromASCII("Array Iterator")), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
-
+    m_arrayPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state.context()->staticStrings().unscopables),
+                                                       ObjectPropertyDescriptor(blackList));
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: `This property has the attributes { [[Writable]]: false, [[Enumerable]]: false, [[Configurable]]: true }.`

**Comment by kisbg** [comment]
Ready for review !

File: src/runtime/GlobalObjectBuiltinArray.cpp (hunk 10 above)
Comment (rerobika) [reply src/runtime/GlobalObjectBuiltinArray.cpp]: I think this issue is still present.

**Comment by kisbg** [comment]
@rerobika  Thank you for the review and for the advice !

File: test/es2015/symbol-unscopables.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+object1[Symbol.unscopables] = {
+ property1: true,
+ property2: false
+};
+try {
+    with (object1) {
+     print(property1);
+     // expected output: Error: property1 is not defined
+    }
+
```
Comment (clover2123) [inline test/es2015/symbol-unscopables.js]: Please add a check code at this point that confirms an exception is occured above, such as inserting `assert(false)` at here. If exception does not occurred, this code will be excuted as normal.

**Comment by clover2123** [comment]
@kisbg Could you add additional TC code that checks `Array.prototype[@@unscopables]` property?

File: src/runtime/EnvironmentRecord.h [hunk 12]
```diff
@@ -216,6 +219,10 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(name));
         if (result.hasValue()) {
+            Value unscopables = m_bindingObject->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, m_bindingObject);
+            if (unscopables.isObject() == true && unscopables.asObject()->get(state, ObjectPropertyName(state, name)).value(state, unscopables).toBoolean(state) == true) {
+                return GetBindingValueResult();
```
Comment (clover2123) [inline src/runtime/EnvironmentRecord.h]: `@@unscopables` is rarely used, so how about adding `UNLIKELY` macro for this if statement?

File: src/runtime/EnvironmentRecord.h [hunk 13]
```diff
@@ -226,6 +233,10 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(atomicName));
         if (result.hasValue()) {
+            Value unscopables = m_bindingObject->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, m_bindingObject);
+            if (unscopables.isObject() == true && unscopables.asObject()->get(state, ObjectPropertyName(state, atomicName)).value(state, unscopables).toBoolean(state) == true) {
+                return BindingSlot(this, SIZE_MAX);
```
Comment (clover2123) [inline src/runtime/EnvironmentRecord.h]: ditto

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 14]
```diff
@@ -1751,6 +1751,16 @@ void GlobalObject::installArray(ExecutionState& state)
...
 
+    Object* blackList = new Object(state);
+    blackList->setPrototype(state, Value(Value::Null));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().copyWithin), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().entries), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().fill), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().find), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().findIndex), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().keys), ObjectPropertyDescriptor(Value(true)));
+    blackList->defineOwnProperty(state, ObjectPropertyName(state.context()->staticStrings().values), ObjectPropertyDescriptor(Value(true)));
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Each property of blacklist should have ` PropertyDescriptor{[[Value]]: V, [[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: true}`, but default parameter of [ObjectPropertyDescriptor](https://github.com/pando-project/escargot/blob/7f7f317340793ed2b593c8537c9c98ecaa1f58e8/src/runtime/Object.h#L233) would set the `NotPresent` attibute which is wrong.

Comment (clover2123) [reply src/runtime/GlobalObjectBuiltinArray.cpp]: Please fix this code as follow.
`ObjectPropertyDescriptor(Value(true), ObjectPropertyDescriptor::AllPresent)`

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 15]
```diff
@@ -1835,7 +1845,7 @@ void GlobalObject::installArray(ExecutionState& state)
                                                                ObjectPropertyDescriptor(new FunctionObject(state, NativeFunctionInfo(state.context()->staticStrings().next, builtinArrayIteratorNext, 0, nullptr, NativeFunctionInfo::Strict)), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::WritablePresent | ObjectPropertyDescriptor::ConfigurablePresent)));
     m_arrayIteratorPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state, Value(state.context()->vmInstance()->globalSymbols().toStringTag)),
                                                                ObjectPropertyDescriptor(Value(String::fromASCII("Array Iterator")), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
-
+    m_arrayPrototype->defineOwnPropertyThrowsException(state, ObjectPropertyName(state.context()->staticStrings().unscopables), ObjectPropertyDescriptor(blackList, (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: This is a minor suggestion. For code consistency, could you move upward this line along with other `m_arrayPrototype->defineOwnPropertyThrowsException` lines?

**Comment by kisbg** [comment]
@clover2123  @rerobika  Thank you for the review !

File: src/runtime/EnvironmentRecord.h (hunk 12 above)
Comment (rerobika) [reply src/runtime/EnvironmentRecord.h]: #296 has landed, so please update the coding style according to it.

File: src/runtime/EnvironmentRecord.h [hunk 16]
```diff
@@ -247,6 +258,7 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
         return m_bindingObject->deleteOwnProperty(state, name);
     }
 
+
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: Unnecessary new line

File: src/runtime/StaticStrings.cpp [hunk 17]
```diff
@@ -94,6 +94,7 @@ void StaticStrings::initStaticStrings(AtomicStringMap* atomicStringMap)
     getSymbolSpecies.init(atomicStringMap, "get [Symbol.species]", strlen("get [Symbol.species]"));
     getSymbolSearch.init(atomicStringMap, "get [Symbol.search]", strlen("get [Symbol.search]"));
     getSymbolToStringTag.init(atomicStringMap, "get [Symbol.toStringTag]", strlen("get [Symbol.toStringTag]"));
+    getSymbolUnscopables.init(atomicStringMap, "get [Symbol.unscopables]", strlen("get [Symbol.unscopables]"));
     symbolSplit.init(atomicStringMap, "[Symbol.split]", strlen("[Symbol.split]"));
```
Comment (clover2123) [inline src/runtime/StaticStrings.cpp]: `getSymbolUnscopables` seems unnecessary. Other looks fine to me.

Comment (kisbg) [reply src/runtime/StaticStrings.cpp]: Yes. You are right. I fixed it.

File: src/runtime/EnvironmentRecord.h [hunk 18]
```diff
@@ -222,6 +224,10 @@ class ObjectEnvironmentRecord : public EnvironmentRecord {
     {
         auto result = m_bindingObject->get(state, ObjectPropertyName(name));
         if (result.hasValue()) {
+            Value unscopables = m_bindingObject->getOwnProperty(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().unscopables)).value(state, m_bindingObject);
```
Comment (rerobika) [inline src/runtime/EnvironmentRecord.h]: You must use `[[Get]]` here.

**Comment by rerobika** [comment]
@clover2123 The patch is ready for the review. Unfortunately, due to Proxy objects in the prototype chain we cannot avoid calling [[Has]] internal method during lexical reference lookup. It may burdens the performance, but it finally properly follows the standard.

File: src/runtime/Object.cpp [hunk 19]
```diff
@@ -779,20 +784,17 @@ ValueVector Object::ownPropertyKeys(ExecutionState& state)
 // https://www.ecma-international.org/ecma-262/6.0/#sec-ordinary-object-internal-methods-and-internal-slots-get-p-receiver
 ObjectGetResult Object::get(ExecutionState& state, const ObjectPropertyName& propertyName)
 {
-    Object* temp = this;
-    auto desc = this->getOwnProperty(state, propertyName);
-    while (true) {
-        if (!desc.hasValue()) {
```
Comment (clover2123) [inline src/runtime/Object.cpp]: Why did you fix this `Object::get` method? Is there any issue related to the feature?

File: src/runtime/Object.cpp [hunk 20]
```diff
@@ -519,6 +519,11 @@ ObjectGetResult Object::getOwnProperty(ExecutionState& state, const ObjectProper
     return ObjectGetResult();
 }
 
+ObjectGetResult Object::getOwnProperty(ExecutionState& state, const ObjectPropertyName& propertyName) ESCARGOT_OBJECT_SUBCLASS_MUST_REDEFINE
+{
+    return this->ordinaryGetOwnProperty(state, propertyName);
```
Comment (clover2123) [inline src/runtime/Object.cpp]: Changing `Object::getOwnProperty' to invoke `Object::ordinaryGetOwnProperty` in itself may affect the performance slowdown because of additional function call. IMO it's better to merge these two methods. How about your opinion?

File: src/runtime/Object.cpp [hunk 21]
```diff
@@ -861,21 +863,21 @@ bool Object::set(ExecutionState& state, const ObjectPropertyName& propertyName,
     return true;
 }
 
-bool Object::ordinaryHasProperty(ExecutionState& state, Object* object, const ObjectPropertyName& propertyName)
+bool Object::ordinaryHasProperty(ExecutionState& state, const ObjectPropertyName& propertyName)
 {
```
Comment (clover2123) [inline src/runtime/Object.cpp]: According to the [spec](https://www.ecma-international.org/ecma-262/6.0/#sec-ordinaryhasproperty), `Object::ordinaryHasProperty` method has two parameters (`OrdinaryHasProperty (O, P)`) which are Object O and PropertyKey P. To match the spec, object parameter seems to be maintained.

File: src/runtime/StringObject.cpp [hunk 22]
```diff
@@ -48,6 +48,11 @@ void* StringObject::operator new(size_t size)
     return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
 }
 
+bool StringObject::hasProperty(ExecutionState& state, const ObjectPropertyName& P) ESCARGOT_OBJECT_SUBCLASS_MUST_REDEFINE
+{
+    auto r = getOwnProperty(state, P);
```
Comment (clover2123) [inline src/runtime/StringObject.cpp]: Why did you add this `hasProperty` method for StringObject and others? According to the ES2015 spec, it seems that there are only 2 defined `hasProperty` methods which are for `Object` and `Proxy` only

**Comment by clover2123** [comment]
@rerobika 
As you said, `hasProperty` is not fully used accoding to the standard. We've noticed this issue and we will fix it soon. So, would you please hold this patch for a while until the `hasProperty` is updated?

File: src/runtime/StringObject.cpp (hunk 22 above)
Comment (rerobika) [reply src/runtime/StringObject.cpp]: It's all about overriding and fast cases. Since each special object implements it's own GetOwnPropety routine while overriding the Object::getOwnProperty, IMO we should do the same for hasProperty as well.

Comment (rerobika) [reply src/runtime/StringObject.cpp]: It's all about overriding and fast cases. Since each special object implements it's own GetOwnPropety routine while overriding the Object::getOwnProperty, IMO we should do the same for hasProperty as well.

File: src/runtime/Object.cpp (hunk 21 above)
Comment (rerobika) [reply src/runtime/Object.cpp]: Indeed, but the `O` always represents the object itself to check whether it has property or not. So at this point the `O` can be substituted with the `this` argument.

File: src/runtime/Object.cpp (hunk 20 above)
Comment (rerobika) [reply src/runtime/Object.cpp]: Sure, thanks for the remind I also wanted to merge these functions.

File: src/runtime/Object.cpp (hunk 19 above)
Comment (rerobika) [reply src/runtime/Object.cpp]: This change is a reverts the function to it's original state. The reason is that in ES5.1 we could simply iterate through the prototype chain with a single while loop, since there was only one [[Get]] internal method. That's why the recursive algorithm of the standard could be changed into the iterative one. However, in ES6 there are proxy objects. When you call a `obj->get` for a simple object who has a proxy object in  its prototype chain, you must invoke the ProxyObject::get during the lookup, as the standard says. Currently we use iterative implementation, which invokes the ProxyObject::getOwnProperty during the lookup. I hope you see my point.

**Comment by rerobika** [comment]
@clover2123
I know that this solution is not the most optimal way to do, but at first I tried to achieve correctness, which can be optimized later.

File: src/runtime/Object.cpp (hunk 19 above)
Comment (clover2123) [reply src/runtime/Object.cpp]: I see. It looks inevitable. Maybe we could optimize `Object::get` later if possible.

**Comment by clover2123** [comment]
@rerobika 
`hasProperty` has been revised at #396, so please rebase this patch to the master branch.
Adding specialized `hasProperty` method for each Object looks good to me, but it would be great if you separate this patch into 2 PRs because this patch now includes 2 main updates which are implementing @@unscopable symbol and some additions related to `hasProperty` method.

**Comment by rerobika** [comment]
@clover2123 Sure, thanks for the info, I'll update the PR soon.

File: src/runtime/Object.cpp (hunk 19 above)
Comment (rerobika) [reply src/runtime/Object.cpp]: I think I found a solution for not using recursion so probably i can resolve this problem.

**Comment by rerobika** [comment]
@clover2123 I'll update the PR after #405 has landed.

**Comment by clover2123** [comment]
@rerobika #405 has been landed. Thanks:)

**Comment by rerobika** [comment]
@clover2123 Please check #407 :)

**Comment by rerobika** [comment]
@clover2123 I've updated the PR
